atomix/cluster/src/main/java/io/atomix/raft/impl/ReconfigurationHelper.java

```diff
@@ -166,6 +170,170 @@ private void leaveInternal(final CompletableFuture<Void> future) {
+   *      |                Commit new config|            AppendEntry              |On commitIndex update               |
+   *      |                                 |------------------------------------>|Commit new config                   |
+   *      |                                 |                                     |                                    |
+   *      |                                 |                                     |      Poll/Vote                     |
+   *      |                                 |                                     |<-----------------------------------|
+   *      |                                 |                                     |  Reject because log not uptodate   |
+   *      |                                 |                                     |----------------------------------->|
+   *      |                                 |                                     |                                    |
+   * </pre>
+   */
+  public CompletableFuture<Void> forceConfigure(final Collection<MemberId> newMembersIds) {
+    final CompletableFuture<Void> future = new CompletableFuture<>();
+
+    threadContext.execute(() -> triggerForceConfigure(newMembersIds, future));
+    return future;
+  }
+
+  private void triggerForceConfigure(
+      final Collection<MemberId> newMembersIds, final CompletableFuture<Void> future) {
+    final var currentConfiguration = raftContext.getCluster().getConfiguration();
+    final Collection<RaftMember> newMembers =
+        newMembersIds.stream()
+            // We can also take the type of the member as input. But so far we do not support
+            // PASSIVE members yet. So we can safely assume they are all ACTIVE.
+            .map(m -> new DefaultRaftMember(m, Type.ACTIVE, Instant.now()))
+            .collect(Collectors.toSet());
+
+    if (!currentConfiguration.force()) {
+      // No need to overwrite if it is already in force configure and this is a retry
+      if (raftContext.getRaftRole().role() == Role.LEADER) {
+        // Optimization: If the current configuration is already the same as new forced, we
+        // can skip reconfiguring. It is most likely a retry of a previous force request,
+        // which was interpreted as failure because of a request timeout.
+        raftContext.transition(Role.FOLLOWER);
+      }
+
+      logger.info(
+          "Current configuration is '{}'. Forcing configuration with members '{}'",
+          currentConfiguration,
+          newMembers);
+      final var newConfiguration =
+          new Configuration(
+              currentConfiguration.index() + 1,
+              raftContext.getTerm(),
+              Instant.now().toEpochMilli(),
+              newMembers,
+              Set.of(),
+              true);
+
+      raftContext.getCluster().configure(newConfiguration);
+    } else if (!(currentConfiguration.allMembers().containsAll(newMembers)
```

Comment:
:upside_down_face: I think you can just use the set `equals` method here. It checks already if the set contains all elements of the other, and if they're the same size:

```java
    public boolean equals(Object o) {
        if (o == this)
            return true;

        if (!(o instanceof Set))
            return false;
        Collection<?> c = (Collection<?>) o;
        if (c.size() != size())
            return false;
        try {
            return containsAll(c);
        } catch (ClassCastException | NullPointerException unused) {
            return false;
        }
    }
```

Reply:
:upside_down_face: Also, since `currentConfiguration.allMembers()` creates a new set every time, I would just fix it to a variable once here and reuse it everywhere else, so we're always looking at the same data set.

Reply:
Ah.. make sense. I thought it was not possible to guarantee that both are sets. But just realized that newMembers are created in the same method. :facepalm: